### DISCUSSION TITLE
Authenticate with AWSCognito seamlessly in web application

### DISCUSSION THREAD

**Discussion by maya-harel** [apache/superset#19012] [Q&A / Help]
I want to integrate Apache Superset into my web application (not a single dashboard, but the whole application).

I am running Superset on an EKS cluster and configuring it using the `helm` chart from this repository.
In my application, I have a Cognito access token saved in the application cookie (also in local storage in the browser). I am trying to find a way to configure the OAuth flow to use the token already saved in order to authenticate, and have this be seamless in the app, (without having to click the "Sign in with AWSCognito" button). 
Or in other words, create an SSO from my application to Superset

Is there a way to configure this?
Or maybe is there an endpoint I can call from my client (with the token) that will sign in and redirect me to the Superset app?

These are my configurations in the helm chart:

`configOverrides:
   enable_oauth: |
     from flask_appbuilder.security.manager import (AUTH_DB, AUTH_OAUTH)
     AUTH_TYPE = AUTH_OAUTH

     OAUTH_PROVIDERS = [
         {
                 "name": "awscognito",
                 "whitelist": [os.getenv('COGNITO_URL'), os.getenv('APP_URL'), f"{os.getenv('APP_URL')}/login"],
                 "icon": "fa-amazon",
                 "token_key": "access_token",
                 "enable_sso": True,
                 "remote_app": {
                     "client_id": os.getenv('CLIENT_ID'),
                     "token_key": "access_token",
                     "api_base_url": os.getenv('COGNITO_URL'),
                     "request_token_url": None,
                     "authorize_url": f"{os.getenv('COGNITO_URL')}/oauth2/authorize?client_id={os.getenv('CLIENT_ID')}&redirect_uri={os.getenv('APP_URL')}&scope=openid&response_type=token"
                     "access_token_url": f"{os.getenv('COGNITO_URL')}/oauth2/token?grant_type=refresh_token&client_id={os.getenv('CLIENT_ID')}&Content-Type='application/x-www-form-urlencoded'",

                 }
             }
     ]
     # Map Authlib roles to superset roles
     AUTH_ROLE_ADMIN = 'Admin'
     AUTH_ROLE_PUBLIC = 'Public'
     # Will allow user self registration, allowing to create Flask users from Authorized User
     AUTH_USER_REGISTRATION = True
     # The default user self registration role
     AUTH_USER_REGISTRATION_ROLE = 'Gamma'`

**Comment by srinify** [comment]
Hey @maya-harel moving to Github Discussions as this isn't a bug :) Cheers!

**Reply by pratikpparikh** [reply]
@maya-harel did you achieve this successfully? if so any documentation that you share for community to use?

**Reply by andreloc** [reply]
@maya-harel were you able to configure it properly? I was wondering if you used this strategy here to authenticate. 
https://aws.amazon.com/blogs/containers/how-to-use-application-load-balancer-and-amazon-cognito-to-authenticate-users-for-your-kubernetes-web-apps/

**Comment by gauravpareekk** [comment]
Hi,

was anyone able able to make it work?

**Comment by nitesh-jaswal** [comment]
Hey, just curious if someone made it work

**Reply by stevetracvc** [reply]
I do it the other way around. I use Superset as the main UI, which redirects the login to Cognito. But then I have iframes embedded in some of my Superset dashboards that load and store a JWT via AWS' react libraries (amplify). The iframes still require authentication, but won't require the user to re-login since they've already authenticated via cognito for Superset.

The pages that are iframes use flask, FastAPI, and react, so this is a pretty specific example that may not help you. I'm pretty sure Apmplify stores the JWT as a gzip base64, or something like that, thus the code in getSession()


I think this is probably the most important part of the library I wrote. I'm not a react or JS/TS developer, so this might not be the best way to do it, but it seems to work well.

This is my 'cognito.tsx' file, in folder 'aws', part of my own private react library.
```
import { useEffect } from 'react';
import { Amplify, Auth, Hub } from 'aws-amplify';

import { CognitoUserType, AmplifyConfigType } from './types';

const pako = require('pako');

function getCookie(cookieName: string) {
  let cookie: { [key: string]: string; } = {};
  document.cookie.split(';').forEach(function(el: string) {
    let [key,value] = el.split('=');
    cookie[key.trim()] = value;
  })
  return cookie[cookieName];
}

export function setup(config: AmplifyConfigType) {
  Amplify.configure(config);
}

export const Login = ({
  onLogin,
} : {
  onLogin: (cognitoUser: CognitoUserType) => void,
}) => {
  useEffect(() => {
      Hub.listen('auth', ({ payload: { event, data } }) => {
        switch (event) {
          case 'signIn':
          case 'cognitoHostedUI':
            console.log('Authenticated...');
            console.log(data);
            onLogin(data);
            break;
          case 'signIn_failure':
          case 'cognitoHostedUI_failure':
            console.log('Error', data);
            break;
        }
      });
    }, []);

  return (
    <div>
      <div style={{width: 100, margin: 'auto', paddingTop: 200}}>
        <button onClick={() => Auth.federatedSignIn()}>
          Log In
        </button>
      </div>
    </div>
  );
};

export const getCognitoUser = () : CognitoUserType | undefined => {
  const val = sessionStorage.getItem('cognitoUser');
  if (val) {
    return JSON.parse(val);
  }
  return undefined;
}

export const getSession = () : string | undefined => {
  const session: string = getCookie('session');
  if (session) {
    // decode from base64, itsdangerous replaces + and /
    const decoded = atob(session.substring(1).replaceAll('-', '+').replaceAll('_', '/').split('.')[0]);

    const decodedArray = Uint8Array.from(decoded, c => c.charCodeAt(0));

    const decomp = JSON.parse(pako.ungzip(decodedArray, { to: 'string' }));

    console.log('decomp', decomp);
    const token = decomp?.oauth?.[" t"]?.[0];
    console.log('session token', token);
    return token;
  }
  return undefined;
}

export const getCognitoToken = () : string | undefined => {
  const cognitoUser = getCognitoUser();
  const token = cognitoUser?.signInUserSession?.accessToken?.jwtToken;
  if (token) {
    return token;
  }
  const session = getSession();
  return session;
}

export const setCognitoUser = (cognitoUser: CognitoUserType) => {
  sessionStorage.setItem('cognitoUser', JSON.stringify(cognitoUser));
}
```



Then I have this in my frontend code:

```
import { CLIENT_ID, COGNITO_BASE } from './secrets';
import { AmplifyConfigType, getCognitoToken } from 'aws';

export const URL_API = "https://MY_FASTAPI_URL";
export const URL_SIGNIN = "https://MY_FLASK_URL";
export const BASE_URL = `${URL_API}/api/v1`;

export const config : AmplifyConfigType = {
    Auth: {
      region: 'YOUR_REGION',
      userPoolId: 'YOUR_POOL_ID',
      userPoolWebClientId: CLIENT_ID,
      mandatorySignIn: true,
      oauth: {
        domain: COGNITO_BASE,
        scope: ['email', 'openid', 'profile'],
        redirectSignIn: URL_SIGNIN,
        redirectSignOut: URL_SIGNIN,
        responseType: 'code',
      },
    },
  };

const buildHeaders = () : Headers => {
  const h = new Headers();
  h.append('Content-Type', 'text/plain');
  h.append('Access-Control-Allow-Origin', '*');

  h.append('Access-Control-Allow-Headers', 'Origin, X-Requested-With, Content-Type, Accept');
  h.append('Access-Control-Allow-Credentials', 'true');
  h.append('Access-Control-Allow-Methods', 'GET, POST');

  const token = getCognitoToken();
  // this object is mutable, and global...so set it here instead
  h.set('Authorization', `Bearer ${token}`);

  return h;
}

export const headers = buildHeaders();

```
